flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/PendingCheckpoint.java

```diff
@@ -115,6 +115,8 @@
 
 	private volatile ScheduledFuture<?> cancellerHandle;
 
+	private final CheckpointFailureManager failureManager;
```

Comment:
Yes, introducing `CheckpointFailureManager` into `PendingCheckpoint` is to avoid scattering calls of `failureManager.handleCheckpointException(exception)` over the coordinator. I accept this suggestion.